Floating-point arithmetic on 386/x87 (32-bit) (#785)
wasmx/fizzy: README.md

If strict compliance is sought with WebAssembly, then the user of Fizzy must ensure to keep the default rounding mode.

### x87 FPU

On the 32-bit Intel i386 architecture an [x87](https://en.wikipedia.org/wiki/X87)-compatible FPU is used by default to perform floating-point operations.
The FPU is claimed to be IEEE-754 compliant, but there is one gotcha. The operations are executed with so-called *internal precision* and the results are rounded to the target precision at the end [[1]](#1).
By default the precision is set to [80-bit extended precision](https://en.wikipedia.org/wiki/Extended_precision) (except for VC++ runtime [[2]](#2)).
Unfortunately, this causes problems for 64-bit double precision operations (`f64.add`, `f64.sub`, `f64.mul`, `f64.div`) as the results may be different than when computed with double precision directly.

> The FPU precision can be dynamically modified by using compiler intrinsics [[1]](#1), but this has similar issues to controlling the rounding mode and there exist no C/C++ standard way of doing so.

Reviewer: ```suggestion
The FPU precision can be dynamically modified by using compiler intrinsics [[1]](#1), but this has similar issues to controlling the rounding mode and there exists no C/C++ standard way of doing so.
```

If strict compliance is sought with WebAssembly, then the user of Fizzy must ensure to keep the default rounding mode.

### x87 FPU

> On Intel i386 architecture (32-bit) the [x87](https://en.wikipedia.org/wiki/X87) FPU is used by default to perform floating-point operations.

Reviewer 2: ```suggestion
On Intel i386 architecture (32-bit) the [x87](https://en.wikipedia.org/wiki/X87) FPU ISA is used by default to perform floating-point operations.
```

or 

```suggestion
On Intel i386 architecture (32-bit) an [x87](https://en.wikipedia.org/wiki/X87)-compatible FPU is used by default to perform floating-point operations.
```

> The FPU is claimed to be IEEE-754 compliant, but there is one gotcha. The operations are executed with so-called _internal precision_ and the results are rounded to the target precision in the end [[1]](#1).

Reviewer 2: ```suggestion
The FPU is claimed to be IEEE-754 compliant, but there is one gotcha. The operations are executed with so-called _internal precision_ and the results are rounded to the target precision at the end [[1]](#1).
```

By default the precision is set to [80-bit extended precision](https://en.wikipedia.org/wiki/Extended_precision) (except for VC++ runtime [[2]](#2)).
Unfortunately, this causes problems for 64-bit double precision operations (`f64.add`, `f64.sub`, `f64.mul`, `f64.div`) — the results may be different than when computed with double precision directly.

The FPU precision can be dynamically modified by using compiler intrinsics [[1]](#1) but this has similar issues to controlling the rounding mode and there exist no C/C++ standard way of doing so.

In the Fizzy implementation we decided it is not worth to fight with the x87 FPU quirks.
Floating-point operations were never our top priorities.
We decided to opt-in for using SSE2 instructions to implement WebAssembly floating-pointing instructions

> also in x86 32-bit builds (this is default for 64-builds). I.e. SSE2 instructions set is required.

Reviewer 2: ```suggestion
also in x86 32-bit builds (this is default for 64-builds). This means for strict WebAssembly compliance the SSE2 instructions set is required.
```

This option is controlled by [`-msse2 -mfpmath=sse`][x86-Options] compiler options and one can always override

> this decision if wants to experiment with x87 FPU.

Reviewer 2: ```suggestion
this decision if wants to experiment with the x87 FPU.
```

> For the record, we also tried to use [`-mpc64`][x86-Options] GCC compiler option which is suppose to set the FPU to 64-bit double precision.

Reviewer 2: ```suggestion
For the record, we also tried to use [`-mpc64`][x86-Options] GCC compiler option which is supposed to set the FPU to 64-bit double precision.
```

> This attempt was unsuccessful for unknown reason.

Reviewer 2: ```suggestion
This attempt was unsuccessful for unknown reasons.
```
